### PR TITLE
Fix: Resolve checkpointing for the Streaming Dataset

### LINKED ISSUE #19122: StreamingDataset state dict not correct
### Bug description

The state dict returned by the streaming dataset always returns index 0. It does not reflect the state of the dataset. In the following code sample, we iterate over the streaming dataset and simulate enough time passing by.

### What version are you seeing the problem on?

master

### How to reproduce the bug

```python
import torch
import time
from lightning.data import StreamingDataset
from lightning.data.streaming.item_loader import TokensLoader
from torch.utils.data import DataLoader
from lightning.fabric import Fabric


def main():
    fabric = Fabric(accelerator="cuda", devices=4)
    fabric.launch()

    dataset = StreamingDataset(
        input_dir="lit-gpt/data/slimpajama/val",
        item_loader=TokensLoader(block_size=2048),
        shuffle=True,
        drop_last=True,
    )
    dataloader = DataLoader(dataset, batch_size=4, pin_memory=True, num_workers=8, drop_last=True)
    dataloader = fabric.setup_dataloaders(dataloader)
    iterator = iter(dataloader)

    for i in range(60):
        next(iterator)
        time.sleep(2)
        fabric.print(i)

    print(dataset.state_dict())


if __name__ == "__main__":
    main()
```

### Error messages and logs

No erros. The output I get is:

```
{'0': {'rank': 0, 'current_epoch': 1, 'input_dir_path': '/cache/chunks/5587867a69455f09b8d5564b7b4ff7b5/0', 'input_dir_url': 's3://tiny-llama-template/slimpajama/val', 'item_loader': {'block_size': 2048}, 'drop_last': True, 'seed': 42, 'checkpoint_interval': 60, 'chunk_index': 0, 'global_index': 0, 'index': 0, 'world_size': 4, 'num_workers': 8, 'shuffle': True}}
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0): Master (commit 4d154685557881b4ff47083267b5a6328b465d61)
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0): 
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR fixes 2 bugs associated with the StreamingDataset

1) The wrong folder would be passed to the load the state dict

2) The condition to perform checkpointing was wrong.

Screenshot 2023-12-07 at 14 54 36

Fixes #19122

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--19123.org.readthedocs.build/en/19123/

cc @borda

### CODE REVIEW

File: tests/tests_data/streaming/test_dataset.py [hunk 1]
```diff
@@ -556,66 +593,101 @@ def test_resumable_dataset_single_worker(tmpdir):
...
+        input_dir=RemoteDir(cache_dir, data_dir), item_loader=TokensLoader(block_size), shuffle=True
+    )
 
     dataset.current_epoch = 1
 
-    assert dataset.state_dict() == {}
+    # assert dataset.state_dict() == {}
 
-    dataloader = DataLoader(dataset, num_workers=1, batch_size=2, prefetch_factor=1)
+    dataloader = DataLoader(dataset, num_workers=2, batch_size=2, prefetch_factor=1)
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_dataset.py:606]: You might want to change the name of the test now: "test_resumable_dataset_single_worker"

File: tests/tests_data/streaming/test_dataset.py [hunk 2]
```diff
@@ -556,66 +593,101 @@ def test_resumable_dataset_single_worker(tmpdir):
...
 
-    dataset = StreamingDataset(input_dir=str(tmpdir), item_loader=TokensLoader(block_size), shuffle=True)
+    dataset = EmulateS3StreamingDataset(
+        input_dir=RemoteDir(cache_dir, data_dir), item_loader=TokensLoader(block_size), shuffle=True
+    )
 
     dataset.current_epoch = 1
 
-    assert dataset.state_dict() == {}
+    # assert dataset.state_dict() == {}
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_dataset.py:604]:
```suggestion
```

File: tests/tests_data/streaming/test_dataset.py [hunk 3]
```diff
@@ -556,66 +593,101 @@ def test_resumable_dataset_single_worker(tmpdir):
     cache.done()
     cache.merge()
 
-    assert len([f for f in os.listdir(tmpdir) if f.endswith(".bin")]) == 50
+    assert len([f for f in os.listdir(data_dir) if f.endswith(".bin")]) == 50
 
-    dataset = StreamingDataset(input_dir=str(tmpdir), item_loader=TokensLoader(block_size), shuffle=True)
+    dataset = EmulateS3StreamingDataset(
+        input_dir=RemoteDir(cache_dir, data_dir), item_loader=TokensLoader(block_size), shuffle=True
+    )
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_dataset.py:600]: Maybe we should explicitly set a low `checkpoint_interval` here (to avoid flakiness?) I see below we are sleeping 0.1 seconds before we assert the state dict.

File: tests/tests_data/streaming/test_dataset.py [hunk 4]
```diff
@@ -556,66 +593,101 @@ def test_resumable_dataset_single_worker(tmpdir):
...
     state_dict_2 = dataset.state_dict()
-    assert state_dict_2["0"]["chunk_index"] == 3
+
+    assert state_dict_2["0"]["chunk_index"] == 2
+    assert state_dict_2["0"]["global_index"] == 4
     assert state_dict_2["0"]["index"] == 0
 
+    assert state_dict_2["1"]["chunk_index"] == 1
+    assert state_dict_2["1"]["global_index"] == 2
+    assert state_dict_2["1"]["index"] == 0
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_dataset.py:674]: Can you remind me what "index" is and why is it 0 in all cases here?

**Review by awaelchli — APPROVED** [review]

File: tests/tests_data/streaming/test_dataset.py (hunk 1 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py:606]: done !

File: tests/tests_data/streaming/test_dataset.py (hunk 3 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py:600]: This won't help much as it is already triggered by moving across chunks. The sleep are mostly to ensure the workers had enough time to run.

File: tests/tests_data/streaming/test_dataset.py (hunk 4 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py:674]: This is the index within a chunk.

File: tests/tests_data/streaming/test_dataset.py (hunk 2 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py:604]: done .

**Review by ethanwharris — APPROVED** [review]
